misc/clock.ru

```diff
@@ -17,10 +17,23 @@ class TickTock
     # filter. Filter is going to be very implemenation specific so I suggest
     # channel/subject where channel is a subset of a subject or rather a
     # subject without wildcards. I don't think there is a need for a
-    # key/value. A simple single argument should be fine.
+    # 
+    # [Are you suggesting 'subject' as an alternative to 'pattern'? I think 'pattern' is clearer...]
+    # [...or do you mean that 'subject' forces a server to test the string and decide?]
+    # 
+    # 2) Some servers might choose to implement optional extended features.
+    # For example, some servers might support WebSocket encoding semantics
+    # (i.e., text vs. binary op-codes, data compression / encoding) or even
+    # SSE variations (i.e, Base64 encoding for binary data support).
+    # 
+    # To allow for these further extensions, I believe a Hash (or "named arguments")
+    # scheme is probably preferable.
     subscribe channel: :time
+    # It allows for something like:
+    subscribe channel: :time, as: :text
     # vs
-    subscribe('time')
+    ## Removed(?):
+    # subscribe('time')
   end
```

Comment:
By the way, we also need an unsubscribe but that should match the subscribe in regard to arguments.

Reply:
I agree that we need to support `unsubscribe`.

From a performance / C point of view, I would rather have a subscription object returned. Perhaps the subscription object could support a `close` method?

i.e.:

```ruby
s = subscribe pattern: "t*", as: :binary
s.close
```

v.s.

```ruby
subscribe pattern: "t*", as: :binary
unsubscribe pattern: "t*", as: :binary
# or:
#unsubscribe(s)
```

Reply:
The way I was thinking about implementing not having a subscription object is probably faster but in either case the difference would be small. Lets think of use cases and easy of use instead. I'll have to try out some sample code to see how it feels. I'll be flying tomorrow so will play then.

Reply:
> The way I was thinking about implementing not having a subscription object is probably faster...

Sure, I'll follow whatever you decide on this one. I just had each connection maintain a doubly linked list of subscriptions (so I can easily remove one from the list in O(1)), so I guess I imagined everyone using a similar solution 😅

But re-finding the subscription could work just as well (the main storage is a HashMap, so I guess it shouldn't make a huge difference (it's just a slower O(1)).

This will also save the client from storing subscription objects, so it might be a positive thing 👍🏻

Reply:
I gave it some though (while in the shower) 😊 I like you idea of a subscription object. Let’s go with that. I’ll use a C hash though. The more kept in C the better.

Reply:
I'm using [a C HashMap](https://github.com/boazsegev/facil.io/blob/master/lib/facil/core/types/fiobj/fio_hashmap.h) as well, at least for the channels ☺️👍🏻

For the clients I use both a HashMap (in C) and a Doubly Linked list (for each channel).

The client HashMap allows me to search subscriptions by properties and prevent duplicated subscriptions by using reference counting (these are practically the only use cases for the HashMap) and the list lets me quickly remove clients from the channel.